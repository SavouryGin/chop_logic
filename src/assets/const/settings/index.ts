import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
};
