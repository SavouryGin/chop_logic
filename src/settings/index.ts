import { SelectEntity } from 'types';

export const languageOptions: SelectEntity[] = [
  { option: { en: 'English', ru: 'Английский' }, value: 'en' },
  { option: { en: 'Russian', ru: 'Русский' }, value: 'ru' },
];

export const languageStringOptions = ['en', 'ru'];

export const externalLinks: { [key in string]: string } = {
  telegram: 'https://telegram.me/savoury_gin',
  gitHub: 'https://github.com/SavouryGin',
  linkedIn: 'https://www.linkedin.com/in/dmitrii-suroviagin/',
  mail: 'mailto:Dmitrii.Suroviagin@gmail.com',
  facebook: 'https://www.facebook.com/dmitrii.suroviagin',
};
