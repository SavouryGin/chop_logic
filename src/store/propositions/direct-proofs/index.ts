import converter from 'logic/propositions/converter';
import executor from 'logic/propositions/executor';
import replacer from 'logic/propositions/replacer';
import validator from 'logic/propositions/validator';
import { DP_INITIAL_STATE } from './initial-state';
import { DirectProofsTableItem, PropositionsDirectProofsFlag } from './interfaces';
import { LocalText } from 'types';
import { PayloadAction, createSlice } from '@reduxjs/toolkit';

export const propositionsDPSlice = createSlice({
  name: 'Prop-DP',
  initialState: DP_INITIAL_STATE,
  reducers: {
    setUpFlag: (state, action: PayloadAction<{ flag: PropositionsDirectProofsFlag; value: boolean }>) => {
      const { flag, value } = action.payload;
      state.flags[flag] = value;
    },

    setSelectedIds: (state, action: PayloadAction<string[]>) => {
      state.selectedIds = action.payload;
    },

    resetError: (state, _action: PayloadAction) => {
      state.error = null;
    },

    setTableData: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      state.tableData = action.payload;
    },

    setClipboardData: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      state.clipboardData = action.payload;
    },

    setDependentItems: (state, action: PayloadAction<DirectProofsTableItem[]>) => {
      state.dependentItems = action.payload;
    },

    addPremise: (state, action: PayloadAction<string>) => {
      const rawInput = action.payload;
      const expression = converter.convertStringToExpression(rawInput);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const id = crypto.randomUUID();
      const newItem: DirectProofsTableItem = {
        id,
        step,
        rawInput,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'Premise', ru: 'Посылка' },
        dependentOn: null,
      };
      state.tableData = [...state.tableData, newItem];
    },

    deleteSteps: (state, _action: PayloadAction<{ isConfirmed: boolean }>) => {
      return state;
    },

    reiterateStep: (state) => {
      const selectedStep = state.tableData.find((item) => item.id === state.selectedIds[0]);
      if (!selectedStep) {
        return state;
      }
      const index = state.tableData.length + 1;
      const newStep = {
        ...selectedStep,
        id: crypto.randomUUID(),
        step: index,
        comment: { en: `Reiter. ${selectedStep.step}`, ru: `Повтор ${selectedStep.step}` },
      };

      state.selectedIds = [];
      state.tableData = [...state.tableData, newStep];
    },

    createImplication: (state, action: PayloadAction<{ firstVariable: string; secondVariable: string }>) => {
      const { firstVariable, secondVariable } = action.payload;
      const expression = converter.convertToICExpression(firstVariable, secondVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const newItem: DirectProofsTableItem = {
        step,
        id: crypto.randomUUID(),
        rawInput: `${firstVariable}, ${secondVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'IC', ru: 'ВИ' },
        dependentOn: null,
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    createImplicationDistribution: (
      state,
      action: PayloadAction<{ firstVariable: string; secondVariable: string; thirdVariable: string }>,
    ) => {
      const { firstVariable, secondVariable, thirdVariable } = action.payload;
      const expression = converter.convertToIDExpression(firstVariable, secondVariable, thirdVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const newItem: DirectProofsTableItem = {
        step,
        id: crypto.randomUUID(),
        rawInput: `${firstVariable}, ${secondVariable}, ${thirdVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'ID', ru: 'ДИ' },
        dependentOn: null,
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    createContradictionRealization: (state, action: PayloadAction<{ firstVariable: string; secondVariable: string }>) => {
      const { firstVariable, secondVariable } = action.payload;
      const expression = converter.convertToCRExpression(firstVariable, secondVariable);
      const formula = converter.convertExpressionToFormula(expression);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const newItem: DirectProofsTableItem = {
        step,
        id: crypto.randomUUID(),
        rawInput: `${firstVariable}, ${secondVariable}`,
        expression,
        friendlyExpression,
        formula,
        comment: { en: 'CR', ru: 'СП' },
        dependentOn: null,
      };
      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    eliminateImplication: (state, _action: PayloadAction) => {
      const selectedIds = state.selectedIds;
      const items = state.tableData.filter((item) => selectedIds.includes(item.id));

      if (items.length !== 2) {
        return state;
      }

      const formulas = items.map((item) => item.formula);

      if (!validator.isIEApplicable(formulas[0], formulas[1])) {
        return state;
      }

      const formula = executor.performIE(formulas[0], formulas[1]);
      const expression = converter.convertFormulaToExpression(formula);
      const friendlyExpression = converter.convertFormulaToUserFriendlyExpression(formula);
      const step = state.tableData.length + 1;
      const newItem: DirectProofsTableItem = {
        step,
        expression,
        friendlyExpression,
        formula,
        id: crypto.randomUUID(),
        rawInput: `${items[0].rawInput}, ${items[1].rawInput}`,
        comment: { en: `IE: ${items[0].step}, ${items[1].step}`, ru: `УИ: ${items[0].step}, ${items[1].step}` },
        dependentOn: [items[0].id, items[1].id],
      };

      state.selectedIds = [];
      state.tableData = [...state.tableData, newItem];
    },

    replacePropositionalVariable: (state, action: PayloadAction<{ newVariable: string; oldVariable: string }>) => {
      const { newVariable, oldVariable } = action.payload;
      state.tableData = replacer.replacePropositionalVariableInDPTableItems(state.tableData, newVariable, oldVariable);
    },

    setError: (state, action: PayloadAction<LocalText | null>) => {
      state.error = action.payload;
    },

    exportToXML: (state, _action: PayloadAction<string | undefined>) => {
      return state;
    },

    importFromXML: (state, _action: PayloadAction<{ file: File }>) => {
      return state;
    },

    copySteps: (state, _action: PayloadAction) => {
      const { selectedIds, tableData } = state;
      const selectedItems = tableData.filter((item) => selectedIds.includes(item.id));

      state.clipboardData = selectedItems.map((item) => ({ ...item, id: crypto.randomUUID() }));
    },

    cutSteps: (state, _action: PayloadAction<{ isConfirmed: boolean }>) => {
      return state;
    },

    pasteSteps: (state, _action: PayloadAction) => {
      return state;
    },
  },
});

export const dpActions = { ...propositionsDPSlice.actions };
