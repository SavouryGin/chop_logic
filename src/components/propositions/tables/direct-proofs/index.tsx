import React, { memo, useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'presets/propositions';
import { propositionsDirectProofsActions as actions } from 'store/propositions/direct-proofs/slice';
import { fillerText } from 'texts/propositions';
import { propositionsDirectProofsSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const DirectProofsEditorTable = () => {
  const dispatch = useAppDispatch();
  const tableData = useAppSelector(selectors.getDirectProofsTableData);
  const language = useAppSelector(settingsSelectors.getLanguage);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.getSelectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='direct-proofs-editor__filler'>{fillerText[language]}</div>;

  return (
    <div className='direct-proofs-editor__table'>
      <Table columns={constants.directProofsEditorTableColumns} data={tableData} hasCheckboxColumn passSelectedIds={takeSelectedIds} />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default memo(DirectProofsEditorTable);
