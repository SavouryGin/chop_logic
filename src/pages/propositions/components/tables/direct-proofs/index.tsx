import React, { memo, useEffect, useState } from 'react';
import Table from 'components/table';
import constants from 'pages/propositions/constants';
import texts from 'texts/propositions/elements';
import { dpActions as actions } from 'store/propositions/direct-proofs';
import { dpSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { settingsSelectors } from 'store/settings/selectors';
import { useAppDispatch, useAppSelector } from 'hooks';

const DirectProofsEditorTable = (): React.ReactElement => {
  const dispatch = useAppDispatch();
  const tableData = useAppSelector(selectors.tableData);
  const language = useAppSelector(settingsSelectors.language);
  const [selectedIds, setSelectedIds] = useState<string[]>(useAppSelector(selectors.selectedIds));
  const takeSelectedIds = (ids: string[]) => setSelectedIds(ids);

  useEffect(() => {
    dispatch(actions.setSelectedIds(selectedIds));
  }, [selectedIds]);

  const noStepsFiller = <div className='direct-proofs-editor__filler'>{texts.fillerText[language]}</div>;

  return (
    <div className='direct-proofs-editor__table'>
      <Table
        columns={constants.directProofsEditorTableColumns}
        data={tableData}
        hasCheckboxColumn
        passSelectedIds={takeSelectedIds}
        id='propositions-dp-editor-table'
      />
      {tableData.length === 0 && noStepsFiller}
    </div>
  );
};

export default memo(DirectProofsEditorTable);
