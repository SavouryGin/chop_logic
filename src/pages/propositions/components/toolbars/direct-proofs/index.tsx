import Button from 'components/controls/button';
import React, { memo } from 'react';
import { ButtonID } from 'enums';
import { propositionsDPActions as actions } from 'store/propositions/direct-proofs/slice';
import { propositionsDPSelectors as selectors } from 'store/propositions/direct-proofs/selectors';
import { soundPlayer } from 'helpers/sounds';
import { useAppDispatch, useAppSelector, useIsImplicationEliminationPossible } from 'hooks';

const DirectProofsEditorToolbar = () => {
  const dispatch = useAppDispatch();
  const selectedIds = useAppSelector(selectors.getSelectedIds);
  const tableDataLength = useAppSelector(selectors.getTableDataLength);
  const isReiterationDisabled = selectedIds.length !== 1;
  const isDeleteDisabled = selectedIds.length === 0;
  const isImplicationEliminationEnabled = useIsImplicationEliminationPossible(selectedIds);
  const isReplacerDisabled = tableDataLength === 0;
  const selectedItems = useAppSelector(selectors.getSelectedTableItems);

  const deleteSteps = () => {
    dispatch(actions.deleteSteps({ isConfirmed: false }));
  };

  const reiterateStep = () => {
    dispatch(actions.reiterateStep());
  };

  const openPremise = () => {
    dispatch(actions.setUpFlag({ flag: 'isPremiseOpened', value: true }));
  };

  const openIC = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationCreationOpened', value: true }));
  };

  const openID = () => {
    dispatch(actions.setUpFlag({ flag: 'isImplicationDistributionOpened', value: true }));
  };

  const openCR = () => {
    dispatch(actions.setUpFlag({ flag: 'isContradictionRealizationOpened', value: true }));
  };

  const performIE = () => {
    dispatch(actions.eliminateImplication(selectedItems));
  };

  const openReplacer = () => {
    dispatch(actions.setUpFlag({ flag: 'isReplacerFormOpened', value: true }));
  };

  return (
    <div className='direct-proofs-editor__toolbar'>
      <Button buttonId={ButtonID.Premise} sound={soundPlayer.keyboard} size='large' onClick={openPremise} />
      <Button
        buttonId={ButtonID.Reiteration}
        sound={soundPlayer.keyboard}
        size='large'
        isDisabled={isReiterationDisabled}
        onClick={reiterateStep}
      />
      <Button
        buttonId={ButtonID.Replace}
        sound={soundPlayer.keyboard}
        size='large'
        onClick={openReplacer}
        isDisabled={isReplacerDisabled}
      />
      <Button buttonId={ButtonID.Delete} sound={soundPlayer.keyboard} size='large' isDisabled={isDeleteDisabled} onClick={deleteSteps} />
      <Button buttonId={ButtonID.ImplicationCreation} sound={soundPlayer.slideClick} size='large' onClick={openIC} />
      <Button buttonId={ButtonID.ImplicationDistribution} sound={soundPlayer.slideClick} size='large' onClick={openID} />
      <Button buttonId={ButtonID.ContradictionRealization} sound={soundPlayer.slideClick} size='large' onClick={openCR} />
      <Button
        buttonId={ButtonID.ImplicationElimination}
        sound={soundPlayer.slideClick}
        isDisabled={!isImplicationEliminationEnabled}
        size='large'
        onClick={performIE}
      />
    </div>
  );
};

export default memo(DirectProofsEditorToolbar);
