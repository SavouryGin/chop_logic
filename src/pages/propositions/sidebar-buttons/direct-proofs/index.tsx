import Button from 'components/controls/button';
import React from 'react';
import { ButtonID, Icon } from 'enums';
import { propositionsDPActions as actions } from 'store/propositions/direct-proofs/slice';
import { propositionsDPSelectors as selectors } from 'store/propositions/direct-proofs/selectors';

import { useAppDispatch, useAppSelector } from 'hooks';

const PropositionsDPSidebarButtons = ({ isVisible }: { isVisible: boolean }): React.ReactElement | null => {
  const dispatch = useAppDispatch();
  const tableDataLength = useAppSelector(selectors.getTableDataLength);

  if (!isVisible) {
    return null;
  }

  // const isPDFSaveDisabled = true;
  const isExportToXMLDisabled = !tableDataLength;

  const onClickExportToXML = () => {
    dispatch(actions.exportToXML());
  };

  const openFileInputForm = () => {
    dispatch(actions.setUpFlag({ flag: 'isUserFileFormVisible', value: true }));
  };

  const onCopySteps = () => {
    console.log('Copy');
  };

  const onPasteSteps = () => {
    console.log('Paste');
  };

  return (
    <>
      {/* <li>
        <Button buttonId={ButtonID.SavePDF} icon={Icon.SavePDF} isDisabled={isPDFSaveDisabled} />
        <span className='sidebar__button-span'>Save</span>
      </li> */}
      <li>
        <Button
          buttonId={ButtonID.ExportXML}
          icon={Icon.ExportXML}
          isDisabled={isExportToXMLDisabled}
          onClick={onClickExportToXML}
          size='large'
        />
      </li>
      <li>
        <Button buttonId={ButtonID.ImportXML} icon={Icon.ImportXML} size='large' onClick={openFileInputForm} />
      </li>
      <li>
        <Button buttonId={ButtonID.CopyProof} icon={Icon.ImportXML} size='large' onClick={onCopySteps} />
      </li>
      <li>
        <Button buttonId={ButtonID.PasteProof} icon={Icon.ImportXML} size='large' onClick={onPasteSteps} />
      </li>
    </>
  );
};

export default PropositionsDPSidebarButtons;
