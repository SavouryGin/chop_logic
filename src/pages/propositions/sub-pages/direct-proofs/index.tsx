import Button from 'components/controls/button';
import PropositionsToolbar from 'pages/propositions/toolbar';
import React from 'react';
import TabList from 'components/tab-list';
import formatClass from 'helpers/formatters/format-class-name';
import texts from 'texts/propositions/elements';
import { ButtonID, Icon } from 'enums';
import { propositionsDPActions } from 'store/propositions/direct-proofs';
import { propositionsDPSelectors } from 'store/propositions/direct-proofs/selectors';
import { propositionsDirectProofsTabs } from 'pages/propositions/constants';
import { settingsSelectors } from 'store/settings/selectors';
import { soundPlayer } from 'helpers/sounds';
import { useAppDispatch, useAppSelector } from 'hooks';
import './styles.scss';

const PropositionsDirectProofs = (): React.ReactElement => {
  const language = useAppSelector(settingsSelectors.language);
  const isToolbarOpened = useAppSelector(propositionsDPSelectors.isToolbarOpened);
  const dispatch = useAppDispatch();

  const toggleToolbar = () => {
    dispatch(propositionsDPActions.setUpFlag({ flag: 'isToolbarOpened', value: !isToolbarOpened }));
  };

  const toolBar = (
    <>
      <h2
        className={formatClass(['propositions-direct-proofs__title', Icon.Propositions])}
      >{`${texts.page[language]} > ${texts.direct[language]}`}</h2>
      <Button buttonId={ButtonID.Tools} onClick={toggleToolbar} icon={Icon.Sidebar} sound={soundPlayer.keyboard} view='flat' />
    </>
  );

  return (
    <article className='propositions-direct-proofs'>
      <TabList tabs={propositionsDirectProofsTabs} className='propositions-direct-proofs__tabs' toolBar={toolBar} />
      <PropositionsToolbar isOpened={isToolbarOpened} />
    </article>
  );
};

export default PropositionsDirectProofs;
