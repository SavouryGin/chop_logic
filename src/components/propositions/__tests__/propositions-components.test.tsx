import ContradictionRealizationForm from '../forms/contradiction-realization';
import DirectProofsEditorToolbar from '../toolbars/direct-proofs';
import ImplicationCreationForm from '../forms/implication-creation';
import ImplicationDistributionForm from '../forms/implication-distribution';
import PremiseForm from '../forms/premise';
import React from 'react';
import ReplacerForm from '../forms/replacer';
import renderWithRedux from 'helpers/test-utils/render-with-redux';
import { combineReducers } from '@reduxjs/toolkit';
import { propositionsDirectProofsInitialState, propositionsDirectProofsSlice } from 'store/propositions/direct-proofs/slice';
import { settingsInitialState, settingsSlice } from 'store/settings/slice';

const mockedReducer = combineReducers({
  settings: settingsSlice.reducer,
  propositionsDirectProofs: propositionsDirectProofsSlice.reducer,
});

const mockedState = {
  settings: settingsInitialState,
  propositionsDirectProofs: propositionsDirectProofsInitialState,
};

describe('Propositions Forms component:', () => {
  it('ContradictionRealizationForm matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<ContradictionRealizationForm />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('ImplicationCreationForm matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<ImplicationCreationForm />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('ImplicationDistributionForm matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<ImplicationDistributionForm />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('PremiseForm matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<PremiseForm />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('ReplacerForm matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<ReplacerForm />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });

  it('DirectProofsEditorToolbar matches the snapshot', () => {
    const { asFragment } = renderWithRedux(<DirectProofsEditorToolbar />, mockedReducer, mockedState);
    expect(asFragment()).toMatchSnapshot();
  });
});
