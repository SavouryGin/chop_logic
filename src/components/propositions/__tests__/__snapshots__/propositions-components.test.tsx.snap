// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Propositions Forms component: ContradictionRealizationForm matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="contradiction-realization-form"
  >
    <p>
      Enter values for the meta-variables in the schema:
    </p>
    <p
      class="contradiction-realization-form__formula"
    >
      ( ¬Φ ⇒ Ψ ) ⇒ (( ¬Φ ⇒ ¬Ψ ) ⇒ Φ )
    </p>
    <form
      action="/"
      aria-label="form"
      class="form"
    >
      <div
        class="implication-creation-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_5"
        >
          <span
            class="label__text"
          >
            Φ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_5"
          name="firstVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="implication-creation-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_6"
        >
          <span
            class="label__text"
          >
            Ψ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_6"
          name="secondVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="formula-preview"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="formula-preview"
        >
          <span
            class="label__text"
          >
            Preview
          </span>
        </label>
        <pre
          class="formula formula-preview__formula"
        />
      </div>
      <button
        class="button button_large button_disabled"
        disabled=""
        id="button_id_7"
        title="Apply Settings"
        type="submit"
      >
        <span
          class="button__shadow"
        />
        <span
          class="button__edge"
        />
        <span
          class="button__front chop-icon__check"
          data-testid="button_id_7"
        >
          Apply
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Propositions Forms component: DirectProofsEditorToolbar matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="direct-proofs-editor__toolbar"
  >
    <button
      class="button button_large"
      id="button_id_8"
      title="Enter premise"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_8"
      >
        Premise
      </span>
    </button>
    <button
      class="button button_large button_disabled"
      disabled=""
      id="button_id_9"
      title="Reiterate proof step"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_9"
      >
        Reiteration
      </span>
    </button>
    <button
      class="button button_large button_disabled"
      disabled=""
      id="button_id_10"
      title="Replace symbol"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_10"
      >
        Replace
      </span>
    </button>
    <button
      class="button button_large button_disabled"
      disabled=""
      id="button_id_11"
      title="Delete proof step"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_11"
      >
        Delete
      </span>
    </button>
    <button
      class="button button_large"
      id="button_id_12"
      title="Implication Creation"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_12"
      >
        Implication Creation
      </span>
    </button>
    <button
      class="button button_large"
      id="button_id_13"
      title="Implication Distribution"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_13"
      >
        Implication Distribution
      </span>
    </button>
    <button
      class="button button_large"
      id="button_id_14"
      title="Contradiction Realization"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_14"
      >
        Contradiction Realization
      </span>
    </button>
    <button
      class="button button_large button_disabled"
      disabled=""
      id="button_id_15"
      title="Implication Elimination"
      type="button"
    >
      <span
        class="button__shadow"
      />
      <span
        class="button__edge"
      />
      <span
        class="button__front"
        data-testid="button_id_15"
      >
        Implication Elimination
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Propositions Forms component: ImplicationCreationForm matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="implication-creation-form"
  >
    <p>
      Enter values for the meta-variables in the schema:
    </p>
    <p
      class="implication-creation-form__formula"
    >
      Φ ⇒ ( Ψ ⇒ Φ )
    </p>
    <form
      action="/"
      aria-label="form"
      class="form"
    >
      <div
        class="implication-creation-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_5"
        >
          <span
            class="label__text"
          >
            Φ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_5"
          name="firstVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="implication-creation-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_6"
        >
          <span
            class="label__text"
          >
            Ψ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_6"
          name="secondVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="formula-preview"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="formula-preview"
        >
          <span
            class="label__text"
          >
            Preview
          </span>
        </label>
        <pre
          class="formula formula-preview__formula"
        />
      </div>
      <button
        class="button button_large button_disabled"
        disabled=""
        id="button_id_7"
        title="Apply Settings"
        type="submit"
      >
        <span
          class="button__shadow"
        />
        <span
          class="button__edge"
        />
        <span
          class="button__front chop-icon__check"
          data-testid="button_id_7"
        >
          Apply
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Propositions Forms component: ImplicationDistributionForm matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="implication-distribution-form"
  >
    <p>
      Enter values for the meta-variables in the schema:
    </p>
    <p
      class="implication-distribution-form__formula"
    >
      ( Φ ⇒ ( Ψ ⇒ Χ )) ⇒ (( Φ ⇒ Ψ ) ⇒ ( Φ ⇒ Χ ))
    </p>
    <form
      action="/"
      aria-label="form"
      class="form"
    >
      <div
        class="implication-distribution-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_5"
        >
          <span
            class="label__text"
          >
            Φ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_5"
          name="firstVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="implication-distribution-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_6"
        >
          <span
            class="label__text"
          >
            Ψ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_6"
          name="secondVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="implication-distribution-form__input text-input"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="text_input_id_7"
        >
          <span
            class="label__text"
          >
            Χ :
          </span>
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_7"
          name="thirdVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="formula-preview"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="formula-preview"
        >
          <span
            class="label__text"
          >
            Preview
          </span>
        </label>
        <pre
          class="formula formula-preview__formula"
        />
      </div>
      <button
        class="button button_large button_disabled"
        disabled=""
        id="button_id_7"
        title="Apply Settings"
        type="submit"
      >
        <span
          class="button__shadow"
        />
        <span
          class="button__edge"
        />
        <span
          class="button__front chop-icon__check"
          data-testid="button_id_7"
        >
          Apply
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Propositions Forms component: PremiseForm matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="premise-form"
  >
    <form
      action="/"
      aria-label="form"
      class="form"
    >
      <div
        class="premise-form__input text-input"
      >
        <label
          class="label label_required"
          data-testid="label-test-id"
          for="text_input_id_4"
        >
          <span
            class="label__text"
          >
            Enter a formula:
          </span>
          <abbr
            class="label__asterisk chop-icon__fire"
            title="required"
          />
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_4"
          name="premise"
          type="text"
          value=""
        />
      </div>
      <div
        class="formula-preview"
      >
        <label
          class="label"
          data-testid="label-test-id"
          for="formula-preview"
        >
          <span
            class="label__text"
          >
            Preview
          </span>
        </label>
        <pre
          class="formula formula-preview__formula"
        />
      </div>
      <button
        class="button button_large button_disabled"
        disabled=""
        id="button_id_7"
        title="Apply Settings"
        type="submit"
      >
        <span
          class="button__shadow"
        />
        <span
          class="button__edge"
        />
        <span
          class="button__front chop-icon__check"
          data-testid="button_id_7"
        >
          Apply
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;

exports[`Propositions Forms component: ReplacerForm matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="replacer-form"
  >
    <form
      action="/"
      aria-label="form"
      class="form"
    >
      <div
        class="replacer-form__input text-input"
      >
        <label
          class="label label_required"
          data-testid="label-test-id"
          for="text_input_id_10"
        >
          <span
            class="label__text"
          >
            Variable to replace:
          </span>
          <abbr
            class="label__asterisk chop-icon__fire"
            title="required"
          />
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_10"
          name="oldVariable"
          type="text"
          value=""
        />
      </div>
      <div
        class="replacer-form__input text-input"
      >
        <label
          class="label label_required"
          data-testid="label-test-id"
          for="text_input_id_9"
        >
          <span
            class="label__text"
          >
            New Variable:
          </span>
          <abbr
            class="label__asterisk chop-icon__fire"
            title="required"
          />
        </label>
        <input
          autocomplete="off"
          class="text-input__field"
          id="text_input_id_9"
          name="newVariable"
          type="text"
          value=""
        />
      </div>
      <button
        class="button button_large button_disabled"
        disabled=""
        id="button_id_7"
        title="Apply Settings"
        type="submit"
      >
        <span
          class="button__shadow"
        />
        <span
          class="button__edge"
        />
        <span
          class="button__front chop-icon__check"
          data-testid="button_id_7"
        >
          Apply
        </span>
      </button>
    </form>
  </div>
</DocumentFragment>
`;
